### ISSUE TITLE
linux version.

### ISSUE THREAD

**Issue by rharunzade** [MultiPoolMiner/MultiPoolMiner#104]
It will be great to have a linux version of this app. please, do it.

**Comment by aaronsace** [comment]
If you want to test: https://github.com/PowerShell/PowerShell

I've never used Linux.

**Comment by aaronsace** [comment]
I assume it all runs the same but would need Linux binaries of all the miners?

**Comment by rharunzade** [comment]
it doesn't work in linux powershell.
Here'is log https://pastebin.com/KArTtyvy

**Comment by aaronsace** [comment]
I wonder if it just doesn't support short names 'Sort' is short for 'Sort-Object'

Have you got a newbie Linux install that I could install in Microsoft Myper-V?

**Comment by rharunzade** [comment]
Oh, I've didn't used Hyper-V, but you can try VirtualBox for linux play.
 Or You can try Wubi - install linux as a program in Windows

**Comment by fonyo** [comment]
I've got two new rigs coming in the next couple of weeks (clearly only waiting on cards and mobos). My plan is to install Linux on them. A Linux-MPM would be awesome! I'll do some testing tomorrow.

**Comment by rharunzade** [comment]
@fonyo it will be great.

**Comment by fonyo** [comment]
@aaronsace will I have conflicts if using the latest version of PS v6.0.0 beta2? I can see on the release page that the web request functions and dir/path handling have changed in v6. Is PS v0.5.0 the one or equivalent we use here? I can already see the nightmares incoming if not having the right PS version.

**Comment by aaronsace** [comment]
Scripts in different PowerShell versions are not necessarily forward/backward compatible. 
If you can get it working in v6 beta and still works fine on v5 then that will be great as it will be ready for Linux and v6.

**Comment by rharunzade** [comment]
I think it's not a good idea to use powershell for linux version. There is a native "bash" lang in Linux.

**Comment by aaronsace** [comment]
PowerShell and .net is crossplatform though. I'm not going to rewrite in bash.

**Comment by fonyo** [comment]
@aaronsace Right, I can't do miracles but I'm trying. I managed to get it downloading stuff but as soon as the download finishes it throws "Error: Incorrect command line". Any ideas?

I shall carry on regardless, I really need to dig into your code and familiarise myself with it which I quite enjoy doing while having my 3rd double espresso. :)

**Comment by fonyo** [comment]
Oh it's the 7zip.exe and its implementation :D

**Comment by fonyo** [comment]
Created bin folder manually. If I manually include the extracted folders, the script actually starts running. I see now that this needs a significant rewrite for linux which I am unable to do myself but I can create a trimmed down version that does claymore-dual only and "switches" the dcri values automatically which is the only reason I started doing this.

**Comment by aaronsace** [comment]
I will try this soon and see where the issues lie. Least I know there are people to test it.

**Comment by rharunzade** [comment]
wow, I see you work is going fine. thanx guys and a good luck! you're welcome if you need help or testing!

**Comment by fonyo** [comment]
So it downloads the files then it looks like they get deleted or MPM deletes them assuming a successful 7zip extraction? Not quite sure as MPM does it over and over and after so many tries the downloaded archives just stay for some reason.

I can only test CPU miners as I am running 16.04 in VirtualBox. One other big chunk of work comes from getting all the linux miners. I must be blind but cannot seem to find where you are storing the links to the miners. Please enlighten me.

I haven't mentioned before but I overcome OP's problem by replacing the syntaxes that threw errors (eg. Sort-Object, Compare-Object etc). Not sure why it doesn't like the shorter ones but using longer syntax solves the problem.

**Comment by lee-aaron** [comment]
@fonyo I've started rewriting the ps1 commands to be readable for Linux but I'm coming across issues like the Unblock-File command isn't found. Does it work for you? Perhaps it might just be that PowerShell isn't compatible with 17.04 yet although I managed to get through some errors through downloading the missing (old) libraries.

**Comment by aaronsace** [comment]
This may be useful to skip 'Unblock-File' which is most likely windows specific: 

```
if (Get-Command $cmdName -errorAction SilentlyContinue) {
"$cmdName exists"
}
```

**Comment by lee-aaron** [comment]
I did some more testing. Found out some more syntax errors that I can fix. I ended up creating a bash script to run powershell. Here's a link to what I ran in Ubuntu 17.04 GNOME [https://pastebin.com/83dRXCAQ](url). As fonyo pointed out, the files get downloaded and then deleted. I believe this is due to the fact that the Bin folder doesn't get created as a result of syntax issues. I'll look into fixing a bunch of this.

**Comment by fonyo** [comment]
@lee-aaron Great stuff, mate! Care to share the changed files here so I can compare them with my test branch? Please also include the bash script, maybe I am missing something and you do better.

**Comment by fonyo** [comment]
@aaronsace I cannot seem to find where you are storing the links to the miner archives MPM references to. Can you please enlighten me?

**Comment by aaronsace** [comment]
It's with the files in the miners folder. The Linux version will need different links and different exe files. It will need a variable to state if the miner command line is for Linux or Windows.

**Comment by lee-aaron** [comment]
@fonyo I put the files on my forked version of MPM. It's under a folder for [Linux](https://github.com/lee-aaron/MultiPoolMiner/tree/master/MultiPoolMinerLinux). Everything is there with my bash script and log files if you are interested. I came across Permission Denied errors which indicates that the user doesn't have RWX permissions. So practically we need to change the permissions on each child item which the line with Unblock-File does. There's a linux implementation of it. I think it has to do with sudo chmod.

**Comment by lee-aaron** [comment]
@aaronsace Does the 7z extraction method occur in the Include.ps1 file? 
`Start-Process "7z.exe" "x $FileName -o$(Split-Path $Path)/$FolderName_Old -y -spe" -Wait`
It seems like the [errors](https://github.com/lee-aaron/MultiPoolMiner/blob/master/MultiPoolMinerLinux/Logs/2017-06-15_10-15-59.txt) I'm getting is primarily due to the fact that the Bin folder is missing since zip files don't get extracted. Perhaps I can do further testing with a tar.gz file such as Claymore and see what happens.

**Comment by lee-aaron** [comment]
I did some testing recently. I started basic with unzipping zip files with the unzip command in linux but realized there were other files in different formats. I had to resort to using 7z (Linux version). So far, everything seems to be working as the zip files are being downloaded and then unzipped into the right folders. It's just the Windows miners for experimentation. I'll search up the Linux links and use 7z for them as well. However, I currently am a bit stuck on resolving these errors as I'm practically a newbie to Powershell. While I am a programmer, I'd still like some clarification. Here is a link to my [logs](https://github.com/lee-aaron/MultiPoolMiner/tree/master/MultiPoolMinerLinux/Logs). From my understanding, it looks like Powershell has trouble comparing two objects perhaps of different types? One is null or different type and is being compared to some value? I think resolving the DifferenceObject error will set me on the right path.

**Comment by aaronsace** [comment]
It's not loading the miners correctly. I think it's just saying that 'type' is null. So will need to look into why 'Get-ChildItemContent' is not loading it correctly. 

Have you got this: https://code.visualstudio.com/#alt-downloads

**Comment by aaronsace** [comment]
All the aliases should be changed to the full names then debug from there.

**Comment by lee-aaron** [comment]
I just installed VS and Powershell for it. I'll look into changing the the aliases as they came up as warnings in the editor (ie ForEach should be ForEach-Object etc.). Anything else I should look out for?

**Comment by lee-aaron** [comment]
On a side note, where do you source the miners with the .txt files? I looked through and I didn't see any reference to an Uri. For example, I know Claymore releases a Linux version and I'd like to download that instead of the Windows version. I also am reading from a lot of the other miners that I could practically rebuild their miners for Ubuntu but it will take time.

**Comment by aaronsace** [comment]
If there is no URI listed then you have to download manually and upload with the release.

**Comment by lee-aaron** [comment]
Regarding the Get-ChildItemContent, it's weird. I inserted some print statements to check what values of those variables were and it seems to be fixed? Sometimes it runs fine without those errors and other times it doesn't. I noticed that I also get [Path errors](https://github.com/lee-aaron/MultiPoolMiner/blob/master/MultiPoolMinerLinux/Logs/2017-06-18_11-26-01.txt) which come from my earlier post about the missing Uri. I'll insert some code to skip over things that are null.

**Comment by fonyo** [comment]
Sorry guys, I have been away. The boiling weather makes me go out and enjoy the sun. :) However, I could still do some further testing after applying a few changes to the codebase (bear in mind I'm still no expert at PS) and it seems that pre-downloading the linux miners and replacing the miner references would actually kick off mining. I will post my actuals later on today.

**Comment by fonyo** [comment]
Responding to the last comment of @lee-aaron.

Also bear in mind that Powershell for linux is still in beta. If you just look through the changelog, you'll see how immature it is.

**Comment by lee-aaron** [comment]
@fonyo Yeah. There's a lot of improvements that could be made but do keep me posted if you make progress! I've managed to replace the URI with the Linux version but I am having trouble with renaming the folder as well as calling and running it. I'm thinking about writing this all in bash but it would take a lot of time.

**Comment by aaronsace** [comment]
PowerShell Version 6 won't be beta forever. I've got Linux and PowerShell installed so I will definitely be looking at this once the current version is working well and I've uploaded a release.

**Comment by lee-aaron** [comment]
Well I made a few changes and got things to run once. However it seems that the Start-Process "7z" line in the Include.ps1 is giving me some [trouble again](https://github.com/lee-aaron/MultiPoolMiner/blob/master/MultiPoolMinerLinux/Logs/2017-06-20_16-52-53.txt). It seems like 7z doesn't create the Bin folder for some reason. Would you guys know a workaround? I fixed various other issues such as Sleep. I noticed in the run that actually worked that MPM tried to create a new window but it didn't work and -WindowStyle isn't supported in my version of PowerShell.

**Comment by fonyo** [comment]
> I'm confident I can get this working perfectly on both Windows and Linux. 
@aaronsace You rock, mate.

@lee-aaron Yeah, I would not even attempt to fix the 7z problem as I have never used it before. Not even mentioning the command line usage. Rewriting MPM in bash would, of course, be a huge task, and I don't think it is the right way to go since MPM already proved it can be run in Linux with some workarounds. I expected the final pieces of the two new rigs to arrive 2 days ago. Well, they didn't. I cannot experiment and help with linux stuff until I actually have a non-virtual linux with real cards to work with. Please bear with me (and the ridiculous stock shortages of everything mining related everywhere).

**Comment by lee-aaron** [comment]
@fonyo Actually, I noticed that the provided 7z.exe doesn't extract it for some reason. However, if you download p7zip-full for Linux, it automatically extracts it for you but then the problem of creating a new terminal comes up. Not sure where that code is so I can take a look at it.

**Comment by fonyo** [comment]
TL;DR: 

Well, I live in London and I ordered 8 rx 580 cards from the biggest IT retailer in the UK, ebuyer. They completely screwed me over by cancelling my order as they did not receive enough stock and apparently they did not want customers to wait any longer. They broke the law in terms of consumer rights but I will not pursue it as my time is more valuable. I would definitely win the case, though. So before the nvidia craze started I managed to order 6 gtx 1070 cards. Now you can't get those either or only at ridiculously inflated prices. I am still expecting 2 motherboards, 1 psu, and 18 risers (don't ask) and a 3-way pci-e 1x splitter. Speaking of risers, registered UK limited company on eBay stated their location as London and of course it's a trick to earn more sales and they are sending the product from China. Very nasty trick. I wanted to order from someone in the UK who delivers within days NOT weeks. So the current situation is just crazy. Price of all mining related equipment is inflated or products are simply out of stock. I repurposed an old server motherboard temporarily and I could hook up two Asus Strix 1070 OC cards with ribbon risers, running Win 10 and MPM with my extreme benchmarks that I took even further with NVIDIA. 500 Sol/s per card, I got it very stable today. I'm gonna spend a whole day tomorrow on benchmarking ethash to achieve 30+ MH/s plus 800+ Sia per card. After that's done I will be able to experiment again with Linux and probably sort out the missing Vertcoin stuff for MPM. I wrote a config generator in PHP for sgminer/Vertcoin 2 years ago so setting this up for NVIDIA/MPM should be a breeze! Vertcoin community is also funding someone to create an optimised AMD kernel for Vertcoin. When it is ready, we will see Vertcoin on top of the AMD profitability list as well so this will be even more relevant by then. 

I've got all sorts of AMD cards going back as far as HD6000 series but I only have the 6x1070 from nvidia (4x MSI gaming OC, 2x Asus Strix OC) so I'm very new to nvidia mining. I'm learning as I go.

I truly believe the ultimate mining rig with regards to speed, profit and stability will be achieved with ubuntu/MPM and I don't want to miss out on being a contributor or shaping it as I'm planning to become a full-time miner.

I am not sure why I wrote down my "whole life story" here but I want to assure you guys that I'm here and will be contributing as much as I possibly can. Maybe I just needed to vent. 

/* rant over */

**Comment by raz3671** [comment]
That's great, fonyo.  I'm currently using Win10 for it and it's fine, but I'd love it if there was a solid release for Ubuntu with the same no-sweat functionality.

**Comment by lee-aaron** [comment]
So I've fixed a few problems and have gotten MPM to run on Ubuntu. But I'm still coming across some issues and not sure really how I should approach them.

- Creating a new terminal in Include.ps1
- Rebuilding Linux versions of miners with source code (perhaps will fix auto fail Cryptonight-Nvidia miner)
- Some [clear error (related to Clear-Host in MPM.ps1?)](https://github.com/lee-aaron/MultiPoolMiner/tree/master/MultiPoolMinerLinux/Logs)

Thanks for your help!

**Comment by fonyo** [comment]
Are you using Wine to run the miners? Wouldn't that lower the hashrate of miners if using emulation to run .exe on linux? Very curious.

**Comment by lee-aaron** [comment]
I'm literally just running the shell script on a terminal. Nothing else. I ran it in VS Code to see what issues came up. Quite a few things actually. I think I actually cannot use the exe files on Linux and I do get permission denied errors. The error log will be in my forked repo. Perhaps it's best to rebuild the miners. I'll test it with just one nvidia and one CPU.

**Comment by fonyo** [comment]
Yeah, linux needs linux miners, don't think miner sw would appreciate emulation, although, I was testing pci-e splitters today and I'm amazed how well they work (4 cards at full speed on a single pci-e 1x slot!). It feels like usb, you could even daisy-chain them, until you hit the pci-e lane limit of the processor (or processor/chipset in amd's case). So who knows, mining is just non-complex calculation. If emulator can push the work real-time, it might actually work without noticeable performance degradation. Of course, these are just assumptions.

Bear in mind, powershell (very sadly) is not my area. So I tinker and tinker with these things until they work. I still haven't had time to do much linux stuff but at least I gathered some info on what the best drivers are. AMD is fine but NVIDIA cards also have to overcome of the overclocking difficulties. I don't really want to use smOS, it works amazingly well but it is very limited, doesn't do profit-switching and most likely insecure.

I believe 7z is open-source, free and cross-platform. That will be the way to go with the native powershell MPM. Please do correct me if I'm wrong.

**Comment by lee-aaron** [comment]
@fonyo @aaronsace I managed to rebuild Jayddee's CPU miner but I made another shell script for that. I also changed the Paths in the miners' individual file so it would execute the program. However, it ended up causing more problems and I put it in my log folder. I still am trying to figure out where I should be creating a new terminal and what commands I should run. I'm pretty sure it's the Start-SubProcess function in Include.ps1 and I'll do more testing later. Interestingly, Klaust's miner won't download for me despite having a similar URI to Jayddee's.

**Comment by aaronsace** [comment]
You can experiment launching the miners by going into PowerShell and using 'Start-Process'. Once that works, you know the issue is with MultiPoolMiner.

**Comment by hypnosis-i2p** [comment]
@lee-aaron I don't know about Powershell, but linux commands to run new terminal are:

1. In Mate, it's mate-terminal -e command
2. In Gnome, AFAIU it's gnome-terminal -e command

For other desktop environments, I don't know necessary commands.

**Comment by lee-aaron** [comment]
I did some testing today. I tried what @aaronsace suggested. It did work so I looked to the code about what it does. I noticed that without running the build script in Jayddee's CPU miner, the program runs fine except for the clear errors. It still doesn't open a new window (not sure which command calls a new window Start-Job?). What's interesting is that it actually runs in the background I think because I'm seeing [logs](https://github.com/lee-aaron/MultiPoolMiner/tree/master/MultiPoolMinerLinux/Bin/Ethash-Claymore) from Claymore's miner. 

I also tried implementing some code to run the build scripts after extracting but that didn't work (perhaps someone can point out why it's not working).
```
$BuildScript = Get-ChildItem -Path "$(Split-Path $Path)/$FolderName_New" -Recurse -Filter build.sh -ErrorAction SilentlyContinue -Force
if($BuildScript.Exists)
{
        gnome-terminal --working-directory=~ -e "bash -c $($BuildScript.FullName)"
}
```
I manually ran the build script myself by going into Jayddee's CPU miner directory and running `./build.sh`. 
Next, I got a bunch of processing error data. It seems like PowerShell has issues reading files from Jayddee's miner (different encoding perhaps).

**Comment by aaronsace** [comment]
@lee-aaron So are you working making MultiPoolMiner compatible with PowerShell 6 on Linux?

**Comment by lee-aaron** [comment]
Yep. However I only have Ubuntu 17.04 and PowerShell is finally compatible
with it thru a deb package so I’ll be working on it. Just figuring out
small details and such and will make a new branch for this purpose.

**Comment by aaronsace** [comment]
That's great news!

**Comment by lee-aaron** [comment]
@aaronsace Just hit a road block with converting to Linux. Managed to change batch files to shell scripts but the Include.psm1 has an issue with the Add-Type command. It's a noted [bug](https://github.com/PowerShell/PowerShell/issues/4814) and I tried adding each individual file but it seems that the Cl.cs depends on OpenClException.cs and vice versa. Perhaps someone can quickly write a workaround in the PowerShell github for this or if there is a solution for our purposes?

**Comment by nhinfray** [comment]
Good luck to all working on this much needed port! Really appreciate the help :-D

**Comment by papampi** [comment]
I have written a whattomine.com auto switch for nvOC ubuntu mining os which uses whattomine.com json
If it helps have a look: https://github.com/papampi/WTM_AUTO_SWITCH

nvOC bitcointalk : https://bitcointalk.org/index.php?topic=1854250.0

**Comment by aaronsace** [comment]
@papampi Maybe you can work on adding whattomine for MultiPoolMiner?

**Comment by papampi** [comment]
@aaronsace , would love to help but I'm not so good at windows coding, may be I can help with your linux integration when I get a bit of time.
For now may be you have a look at megaminer :
https://github.com/tutulino/Megaminer/blob/master/Pools/WHATTOMINE_VIRTUAL.ps1

**Comment by lee-aaron** [comment]
@aaronsace 
`$AllPools = @($NewPools) + @(Compare-Object @($NewPools | Select-Object -ExpandProperty Name -Unique) @($AllPools | Select-Object -ExpandProperty Name -Unique) | Where-Object SideIndicator -EQ "=>" | Select-Object -ExpandProperty InputObject | ForEach-Object {$AllPools | Where-Object Name -EQ $_}) | Where-Object {$Algorithm.Count -eq 0 -or (Compare-Object $Algorithm $_.Algorithm -IncludeEqual -ExcludeDifferent | Measure-Object).Count -gt 0}` 
not sure what you mean here. $AllPools.count returns 0 because $AllPools is blank at the creation of the variable. I presume you meant to call $PoolNames? or somehow get the pools the user wants to run? This is in the MultiPoolMiner.ps1. I'm curious as to how it works on windows since it doesn't work the same on Linux.

**Comment by aaronsace** [comment]
@lee-aaron I'm not sure what you are referencing with '$AllPools.Count'...? 

This? 

```
if ($AllPools.Count -eq 0) {
    Write-Warning "No pools available. "
    $Downloader | Receive-Job
    Start-Sleep $Interval
    continue
}
```

**Comment by lee-aaron** [comment]
@aaronsace Specifically 
```@(Compare-Object @($NewPools | Select-Object -ExpandProperty Name -Unique) @($AllPools | Select-Object -ExpandProperty Name -Unique) | Where-Object SideIndicator -EQ "<=" | Select-Object -ExpandProperty InputObject | ForEach-Object {$AllPools | Where-Object Name -EQ $_})``` 
in your MPM.ps1 file is blank. I've updated my version of MPM with the changes you made and am trying to get past this $AllPools part. $AllPools is blank when writing to host and also that the segment of code I copied above is blank too. @($NewPools) is working but not sure why it's not saving that to $AllPools. I tried using Compare-Object on it's own and found that the side indicator is actually <= and not =>

**Comment by aaronsace** [comment]
Break it down and test it: 
If all '$AllPools' is '$null' then this should return '@()': 
`@(Compare-Object @($NewPools | Select-Object -ExpandProperty Name -Unique) @($AllPools | Select-Object -ExpandProperty Name -Unique) | Where-Object SideIndicator -EQ "=>" | Select-Object -ExpandProperty InputObject | ForEach-Object {$AllPools | Where-Object Name -EQ $_}) | Where-Object {$Algorithm.Count -eq 0 -or (Compare-Object $Algorithm $_.Algorithm -IncludeEqual -ExcludeDifferent | Measure-Object).Count -gt 0}`

**Comment by lee-aaron** [comment]
The ForEach-Object loop is where that section of code returns a blank space. It's not null when I'm printing it to console.

**Comment by aaronsace** [comment]
@lee-aaron b9e068ce44dac41b72a4aea9eee55a05f5a1bd67

**Comment by aaronsace** [comment]
@lee-aaron Regarding the miners themselves. You'll have to read the instructions on the miners; I'm no Linux expert. I assume there can be some way to automatically download and run them that we can come up with.

**Comment by lee-aaron** [comment]
[2017-12-07_22-59-10.txt](https://github.com/aaronsace/MultiPoolMiner/files/1541782/2017-12-07_22-59-10.txt) Got some errors and stuff. Not sure how to fix some of them especially the opencl.dll one because linux should be taking .so files. I have finals coming up but I've been working on this a few hours each weekend. I'll have to research more into how to auto download and run them. https://github.com/fireice-uk/xmr-stak/blob/master/doc/Linux_deployment.md I think running them will be along the lines of this.

**Comment by aaronsace** [comment]
Forget OpenCL for now but ultimately this C# code will need to work with OpenCL.so.

**Comment by lee-aaron** [comment]
Ah I think I need to change all opencl.dll names to opencl.so in here https://github.com/aaronsace/opencl-sharp/blob/master/opencl-sharp/NativeMethods.cs But I did notice another error 

```
FullyQualifiedErrorId : ParameterArgumentValidationError,Microsoft.PowerShell.Commands.AddMemberCommand
Add-Member : Cannot validate argument on parameter 'NotePropertyName'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
At /home/aaron/Downloads/MultiPoolMiner/MultiPoolMiner.ps1:135 char:119
+ ... elect-Object -Unique | ForEach-Object {$Pools | Add-Member $_ ($AllPo ...
+                                                                ~~
+ CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
+ FullyQualifiedErrorId : ParameterArgumentValidationError,Microsoft.PowerShell.Commands.AddMemberCommand
```

That was before the opencl.dll one. It seems that other users have also come across this.

**Comment by lee-aaron** [comment]
[2017-12-14_15-17-50.txt](https://github.com/aaronsace/MultiPoolMiner/files/1561102/2017-12-14_15-17-50.txt)

I think it seems to run fine now. I merged the recent most commits to master to my linux branch as of now and I'll be just updating the links to get linux versions etc.

**Comment by aaronsace** [comment]
Awesome, sounds like you're getting somewhere.

**Comment by papampi** [comment]
Hi all
So I finished updating nvOC and have some time to start getting my hands dirty with this amazing idea.
I forked @lee-aaron  repo but its very old.
Possible to update it with latest changes or let me know if there is another repo that I should fork and start from there.
Thanks.

**Comment by lee-aaron** [comment]
@papampi I made a linux branch on my forked repo of MPM. https://github.com/lee-aaron/MultiPoolMiner/tree/linux

**Comment by papampi** [comment]
@lee-aaron 
Thanks just forked it and will start fooling around to get whats what 
The miner section was what I was thinking to ask how you think you gonna handle it ...

I think for close source miners like ewbf, dstm, claymore, ... its easy, we just add download urls and extract them to their locations, the problem is on ccminers and others that need compiling, which will takes big downloads and long compilations.
So we have 2 approaches both need an update/helper bash script.
1st is same as what we do in nvOC, we compile them and make a zip or tar file and put them in a fie server like dropbox then add links to our update bash scripts so all users get them by running the update script.
2nd will be to download them from source and compile on each rig with the same helper/updater bash script.

I'm OK with writing the bash scripts for both ways, but the 2nd will take so long time for every rig to compile. it will be much faster to just download the compiled version and put them in their locations.

**Comment by lee-aaron** [comment]
@papampi I like the first idea. Never really thought of that approach. It sounds like the user just needs to download said zipped/tar file and unzip/tar it then run.

**Comment by papampi** [comment]
I almost have all ccminers, vertminer and many more compiled and ready to download in Dropbox. They don't even need to download and extract, our script will do all for them.
Take a look at my nvOC repo and check nvoc-19-2- update script
I also added a version file in all of the miners folders too so when miners updated my script knows that the miner need to be replaced with new one.
I don't have any RX cards, but have almost all Nvidia miners ready.

**Comment by hbh3** [comment]
Any updates on this?  I'm currently using the Windows version but would love to have a Linux alternative.  I am a Linux shell programmer and I would like to help code and/or test.  I have a rig with 4x NVIDIA cards.

**Comment by lee-aaron** [comment]
@hbh3 If you check out my forked version of MPM, I have already started some shell scripts on the linux branch. However, I would like help on running the other miners. @papampi has provided some good tools to use. It looks like just using curl to fetch said miner, unzip, and run in separate terminal. I am going back to uni soon and will not have much time during the week to work on this project so any help is appreciated!

**Comment by papampi** [comment]
@lee-aaron 
Sorry I've been on a small trip and will come back home in a day or two then I will start working on scripts for miners and if any thing else I can improve

**Comment by nicolasdanelon** [comment]
I have 2 RX470, 1 RX570 and a PNY GTX 1060 6Gb.
I can help with Linux testing on Ubuntu, Arch and Debian. let me know if you need something. I also have experience in bash programming and debugging (:

**Comment by papampi** [comment]
Any one could launch a miner with default start bash commands?
I tried with zpool and miningpoolhib start.sh with no success, manually added dstm miner to  ./Bin/Equihash-DSTM/zm" fixed DSTM.txt path by: 
Path":  "./Bin/Equihash-DSTM/zm 
and still getting    Status: Failed and it can not launch the miner
I think before solving the miner downloads which is the easy part we should solve the launch command. 

`$ ./Start-Zpool.sh 
PowerShell v6.0.0-rc.2
Copyright (c) Microsoft Corporation. All rights reserved.

https://aka.ms/pscore6-docs
Type 'help' to get help.

Transcript started, output file is .\Logs\2018-01-03_12-55-23.txt
The parameter '-WindowStyle' is not supported for the cmdlet 'Start-Process' on this edition of PowerShell.
+ CategoryInfo          : NotImplemented: (:) [Start-Process], NotSupportedException
+ FullyQualifiedErrorId : NotSupportedException,Microsoft.PowerShell.Commands.StartProcessCommand
+ PSComputerName        : localhost
 
   Type: NVIDIA

Miner Algorithm        Speed      BTC/Day BTC/GH/Day Pool 
----- ---------        -----      ------- ---------- ---- 
Dstm  Equihash  Benchmarking Benchmarking  411.82000 Zpool
   Status: Failed

    Speed Active                      Launched Command                                                  
    ----- ------                      -------- -------                                                  
0.00 PH/s 00 Days 00 Hours 00 Minutes Once     Bin/Equihash-DSTM/zm --telemetry --server          
                                               equihash.mine.zpool.ca --port 2142 --user & -username    
                                               papampi -workername multipoolminer -region europe        
                                               -currency btc,usd,eur -type nvidia -poolname zpool       
                                               -algorithm equihash  --pass multipoolminer,c=BTC         


`

**Comment by aaronsace** [comment]
I'm willing to help fund this; if anyone wants to join the MultiPoolMiner organization and believes they can develop/maintain the Linux compatibility then add me on Skype. User: aaronsace

**Comment by hbh3** [comment]
Funding is not really an issue for me. My problem is time. I am looking to do a complete re-write for Linux in Python or Java.  I am also considering storing the profitability data and configuration data in a database such as MariaDB rather than in text files.

Do you have a flowchart for the program(s)?  What about any of the math needed for the profitability calculations?  Those would help me a great deal.  What I may do is just take a few days off work to seriously put my nose to the coding grindstone.

Maybe after I prove I can get this on the road you will be able to put some backing behind the efforts.

**Comment by lee-aaron** [comment]
@hbh3 why are you looking to do a re-write in Python or Java? I presume you mean rewriting the Linux version in those languages?

I think a flowchart would be helpful for me. I am able to successfully call Start-Process "tar" "..." and basically extract the tar.gz folders. I'm just coming across issues such as why there's nothing being printed to console to help me debug. I think either a slack chat or telegram would be helpful for a project such as this rather than simply skype.

**Comment by hbh3** [comment]
Yes, of course, it makes very little sense to try to port the Windows Powershell version to Linux.  A rewrite in Python would also make it multi-platform.

**Comment by lee-aaron** [comment]
@hbh3 I presume you know that there's a Linux Powershell version but aside from that, I very much would like to code in languages I know haha.

**Comment by aaronsace** [comment]
@hbh3 You missed the point; MultiPoolMiner already 'runs' on Linux. The funding is for someone's regular/consistent time to fix the Linux compatibility issues and keep the miners update and complied correctly.

**Comment by hbh3** [comment]
Yes, it runs, but isn’t it in an Alpha state with several bugs?

**Comment by aaronsace** [comment]
@hbh3 MultiPoolMiner 'runs' in PowerShell no matter what OS you have it installed on. It doesn't 'work' and is why I want someone to join the MultiPoolMiner organization.

**Comment by hbh3** [comment]
Understood. :)

**Comment by lee-aaron** [comment]
I don't think I'm the right person to join. I can chip in a few hours each week and do small work but I got uni and I think someone else is better suited for the job to get it running faster.

**Comment by aaronsace** [comment]
@lee-aaron Yes, don't get distracted from your education.

**Comment by hbh3** [comment]
I’ve been trying out all the miner suites, and I keep coming back to MPM as the best and most trusted. I’d like to give development of the Linux version a whirl on Powershell. There are a thousand ways to program everything so I’m sure we can get around any bugs.

**Comment by bviktor** [comment]
I definitely want to have a look at this once I actually have a workstation to spare.

**Comment by papampi** [comment]
Am I missing some thing here?
You guys saying it runs on Linux but I can't get it to launch miner, and I have no knowledge of powershell so can any one help me launch a miner first then I can start work on rest of miner compile and download scripts.

**Comment by papampi** [comment]
Sorry, wrong click

**Comment by aaronsace** [comment]
@papampi hbh3 is now working on a Linux version. This this the branch to follow should you want to keep up on the progress: 
https://github.com/MultiPoolMiner/MultiPoolMiner/tree/Linux

**Comment by lee-aaron** [comment]
I actually made some progress with conversion of batch scripts to shell scripts. Should I make a request to merge into Linux branch? Regarding to get a miner to run on Linux, I can't get ccminerklaust to compile on mine. It keeps saying I'm missing a cuda_runtime_api headerfile. But step wise, I think it would be like

1. wget [link]
2. tar -zxvf [tar.gz] -C [directory]
3. ./autogen.sh ... [other compile/building scripts]
4. run main miner script in new terminal window (with gnome-terminal?)

**Comment by aaronsace** [comment]
@lee-aaron Nice, yes, best to do a pull on that branch then @hbh3 will review/merge it.

**Comment by bviktor** [comment]
TBH I don't quite get it. Why convert from one platform-specific format to another platform-specific format? And then maintain all logic in 2 formats? Why not convert the remaining batches to PowerShell? Or convert it all to Python and be done with it, even :)

**Comment by aaronsace** [comment]
@bviktor PowerShell is open source and cross platform.

**Comment by bviktor** [comment]
Who said it isn't? I even suggested going full PowerShell. I was responding to this:

> I actually made some progress with conversion of batch scripts to shell scripts.

**Comment by aaronsace** [comment]
@bviktor Most likely the start.bat files.

**Comment by vhatzist** [comment]
@lee-aaron 

About below:
> run main miner script in new terminal window (with gnome-terminal?)

How about using screen to run the miner scripts on new screens? This way the Linux implementation of MPM would be desktop-environment agnostic and could even run on Linux machines with no GUI installed. 

Viewing the miner output would be as easy as connecting to the screen that runs it via the CLI.

**Comment by lee-aaron** [comment]
@vhatzist Well I just was providing some sample steps. It is completely flexible to make it usable among all distros.

**Comment by hbh3** [comment]
I'll be investigating setting it up for use with and without window managers.

**Comment by vhatzist** [comment]
@lee-aaron 

> Well I just was providing some sample steps. It is completely flexible to make it usable among all distros.

Do you have any preliminary ideas as to how to support different distros or different desktop environments or even CLI-only machines without a desktop environment? This is why using 'screen' came to mind, or converting MPM to a daemon that will spawn the miners in the background...

**Comment by papampi** [comment]
screen is very good, we use it in nvOC almost for all important scripts (watchdog, miner, temp control ,...)

**Comment by papampi** [comment]
@aaronsace 
Thanks for new linux tree,
I just downloaded it, made some changes for start.sh, miner and some more. but still cant get mpm to launch the miner.
Is this the correct line to start it on linux :
`pwsh -noexit -executionpolicy bypass -file ./MultiPoolMiner.ps1 "&-wallet 1Q24z7gHPDbedkaWDTFqhMF8g7iHMehsCb -username papampi -workername multipoolminer -region europe -currency btc,usd,eur -type nvidia -poolname miningpoolhub -algorithm equihash "`

**Comment by apiontek** [comment]
I am interested to help with this. I've been hacking together my own script in python to fetch pool api data and make decisions on what miners to run, so I have some familiarity with the process. I haven't used powershell much, let alone on linux, but I have 2 linux & 2 win machines doing mining, so I could test things out and contribute as I learn.

**Comment by CharlTruter** [comment]
Hi @hbh3,

Any progress here? Need any help with the conversion to speed things up?

**Comment by hbh3** [comment]
Yes, I've progressed past several hurdles, but there are always new ones to jump over. Thanks for the offer.

**Comment by hbh3** [comment]
Adam, I would not try the linux branch right now as it is not ready.  Still more work to be done.

**Comment by ajpsp5** [comment]
@hbh3 

Do you know of any alternatives for linux based systems?

**Comment by ajpsp5** [comment]
@hbh3 

I end up creating a package for automatically switching to the most profitable algorithm. I hope it helps anyone else that needs an Ethos/Linux alternative.

https://github.com/ajpsp5/ethos-auto-miner

**Comment by sephir69** [comment]
1 hour a day donation seems a bit much to me

**Comment by MaynardMiner** [comment]
I have written a powershell based switching application that was originally a fork of sniffdog miner, but has been completely (about 85%) re-written at this point, which works 100% natively in linux. I didn't even know MPM existed when I started it. I started it, as sniffdog disappeared.

There are a whole bunch of complications, besides just getting powershell to work:

1.) zombie applications. (no process handles to deal with it).
2.) ports fail to close.
3.) api differences for same miner between windows/linux.
4.) process tracking with powershell is horrible in linux.
5.) half the cmdlets in linux exist in powershell (though its getting better).
6.) the moment you make it- everyone wants it to work in every linux distribution imaginable, and expect it to work in their particular flavor right out of the box.
7.) linux mining os developers get very mad at you, very, very, quick.
8.) certain cmdlets may or may not work in linux (straight glitches).

I have written an open sourced miner in powershell that works for linux- SWARM. You are welcome to reference it, but I have re-written most of every function every part of sniffdog miner in order to get it to work, and I would imagine much of mpm would have to be re-written as well.

If you have questions on using powershell linux, let me know. I've become a wizzard at it now. I would assist here, but my project is full-time work at this point.

**Comment by aaronsace** [comment]
@MaynardMiner Well, the last guy who I paid was useless but if you want to join the organization then let me know.

**Comment by MaynardMiner** [comment]
I don't particularly want to join, as my project direction currently deviates/going a different direction than yours, and I would like to stay on top of it (It's a full time job). It's no offense to you- You guys are genius's in your own rights.

But I would like to keep a dialog with you guys, and offer solutions/contributions as you require them. I know it solving problems you may have, only helps me solve mine. Feel free to tag or ask me questions if you see I made something that you haven't implemented yet, and I can explain what I did.

My codebase is open source- You are welcome to reference it as needed.

**Comment by MaynardMiner** [comment]
I'll tell you what, though. Let me finish my next release, because zergpool is back on scene and I have to get them in- And I will try plugging MPM into linux, and see what can be done. If I see ways to make it work, and can formulate an outline to do so, and what needs to be done- I'll message you.

I don't like being the only guy on the linux side. It's a lot of work, as I'm sure you know. It would be nice to come up with a way to bridge projects in some form, even though we are going in different directions.

But I have to plug it in, and calculate time vs. energy to do. I will get back to you.

**Comment by starrbuck68** [comment]
How's that Linux version coming along?
